Support Bucket/Object lock operations

stephenplusplus opened the pull request:
#### To Dos
- Tests
  - [x] System
  - [ ] Unit

#### References
- https://docs.google.com/document/d/1wRpiAxTHZoypR1NLVP97Cj0WCGVKx43Tivso0xXhyEA/edit?ts=5b60f543#
- https://www.googleapis.com/discovery/v1/apis/storage/v1/rest

This introduces new behavior:

# Buckets

- **bucket#lock()**
Lock a previously-defined retention policy. This will prevent changes to the policy.

- **bucket#removeRetentionPeriod()**
Remove an already-existing retention policy from this bucket.

- **bucket#setRetentionPeriod(durationInSeconds)**
Lock all objects contained in the bucket, based on their creation time.

# Files

- **file#getExpirationDate()**
Get a Date object representing the earliest time this file will expire.

---

Additionally, `bucket#getMetadata()`, `bucket#setMetadata()`, `file#getMetadata()`, and `file#setMetadata()` are available if users wish to interact with the raw resource schemas as defined by the API.

frankyn commented on src/bucket.js:
```diff
@@ -1799,6 +1836,34 @@ class Bucket extends common.ServiceObject {
   }
 
+  /**
+   * Remove an already-existing retention policy from this bucket.
```
Only possible if policy is not locked.

frankyn commented on src/bucket.js:
```diff
@@ -1960,6 +2032,51 @@ class Bucket extends common.ServiceObject {
+   * {@link File#removeRetentionPeriod} and {@link File#setRetentionPeriod}. A
+   * locked retention policy cannot be removed or shortened in duration for the
+   * lifetime of the bucket. Attempting to remove or decrease period of a locked
+   * retention policy will result in a `PERMISSION_DENIED` error.
```
Retention policy can be increased when the retention policy is locked.

stephenplusplus replied:
This is actually a copy/paste of what is in the REST docs: https://www.googleapis.com/discovery/v1/apis/storage/v1/rest

I updated ours to say "You can still increase the policy." at the end, although I think it is implied by the previous mentions of "...cannot be removed or shortened..." and "Attempting to remove or decrease...".

Just mentioning in case you want to make the option to increase the policy more pronounced in the REST docs as well.

frankyn replied:
Thanks for the feedback, I'll forward this!

frankyn commented on src/file.js:
```diff
@@ -1827,6 +1871,41 @@ class File extends common.ServiceObject {
+   * released.
+   *
+   * @param {object} [options] Configuration object.
+   * @param {boolean} [options.temporary=false] - Set a temporary hold.
```
Should be set to true in documentation, I think?

stephenplusplus replied:
By default, it is set to false in our code:

```js
if (options.temporary) {
  metadata.temporaryHold = true;
} else {
  metadata.eventBasedHold = true;
}
```

Should we flip it, so that by default, a "file.hold()" creates a temporary hold, as opposed to an event-based hold?

frankyn replied:
My mistake please ignore*

frankyn commented on src/file.js:
```diff
@@ -1827,6 +1871,41 @@ class File extends common.ServiceObject {
+   * @param {SetFileMetadataCallback} [callback] Callback function.
+   * @returns {Promise<SetFileMetadataResponse>}
+   */
+  hold(options, callback) {
```
This is subtle to me, is this considered idiomatic for Node.js?

stephenplusplus replied:
Would you mind elaborating?

frankyn replied:
Sorry, I meant, given that each hold is different, could the surface API have separate methods?

frankyn replied:
I like the idea of having these convenience methods, but it might be overdoing it as a user can update these holds using a setMetadata. What are your thoughts? I'm unsure honestly given the rest of the surface. Looking for input :)

cc: @mohammadnasre

frankyn replied:
So disregard the mention of removing convenience methods. It makes sense here. 

Pending additional feedback from Mohammad.
I think, unifying methods is hard to understand, because at first glance I don't understand what kind of hold the method apply's. 

As a developer, I'd appreciate getting a hint in the name as to what the hold does given bucket lock surfaces two holds.

stephenplusplus replied:
I started fleshing this out into multiple convenient methods. I realized the methods were such simple calls to `setMetadata`, e.g.:

```js
file.setTemporaryHold = function () { file.setMetadata({ temporaryHold: true }) }
file.setEventBasedHold = function () { file.setMetadata({ eventBasedHold: true }) }

file.releaseTemporaryHold = function () { file.setMetadata({ temporaryHold: false }) }
file.releaseEventBasedHold = function () { file.setMetadata({ eventBasedHold: false }) }
```

Seeing that, I started to lean a bit towards skipping the convenience methods, as we're not really adding that much convenience. We could just show an example in the docs for `setMetadata`?

frankyn replied:
I think it add obfuscation, and I appreciate your input. Move forward with documentation instead. This is something we shouldn't over complicate if it doesn't help at all. It's more of an alias that could also help introduce developers to manipulating other metadata.

frankyn replied:
One additional note, system tests should verify how the client library behaves when holds are enabled and attempting to perform an overwrite and delete.

I have additional tests if you'd like to reference tests in Java [PR#3574](https://github.com/GoogleCloudPlatform/google-cloud-java/pull/3574).

frankyn replied:
Friendly ping, @stephenplusplus.

stephenplusplus commented on src/bucket.js:
```diff
@@ -1525,6 +1525,43 @@ class Bucket extends common.ServiceObject {
+   * const storage = require('@google-cloud/storage')();
+   * const bucket = storage.bucket('albums');
+   *
+   * bucket.lock(function(err, apiResponse) {});
```
As far as I can tell, and in my own experience, yes. But, I can't confirm definitively.

frankyn replied:
Sorry, I'll rephrase, `storage.bucket('albums')` always returns bucket metadata unless something on the API changes or user does something different.

The issue with lock is it's a one time thing for every bucket. Once locked it can't be unlocked. A requirement is that it has the metageneration for the bucket to pass as a precondition.

stephenplusplus replied:
Woops, sorry for the misunderstanding.

`storage.bucket('albums')` won't actually return any metadata through our API. The user can call, `storage.bucket('albums').getMetadata()` which will update the `metadata` property on that Bucket's instance.

In the code within `lock()`, we don't check for `bucket.metadata.metageneration` and instead just always do a metadata refresh to get the latest value. Is there a better way to handle that?

frankyn replied:
Thanks, could you make metageneration a parameter instead for retentionLock() and getMetadata() before calling this method to pass in the generation? I'd rather this operation not do a metadata get as well. It defeats the purpose of optimistic concurrency in this case.

stephenplusplus replied:
Done!

stephenplusplus commented:
@frankyn I've added the system tests to verify the hold behavior will block overwrite and delete operations on File objects. I've also removed `file.hold()` and `file.release()` and updated the docs for `file.setMetadata()`.

frankyn commented on src/bucket.ts:
```diff
@@ -1586,6 +1586,46 @@ class Bucket extends ServiceObject {
+   * //-
+   * // If the callback is omitted, we'll return a Promise.
+   * //-
+   * bucket.lock().then(function(data) {
```
Last nit to update sample. Surface LGTM, thanks @stephenplusplus!

stephenplusplus replied:
🤦‍♂️ Updated!
